### PR TITLE
🧠🌴 Fix Braintree 3DS in Dev portal (and much more)

### PR DESCRIPTION
**Pull request by josemigallas**
### Description
[THREESCALE-8967: Adding Braintree SCA card fails](https://issues.redhat.com/browse/THREESCALE-8967)

The 3DS card verification is broken. It happens because we're sending `billingAddress: { region }` as a full name instead of a ISO 3166 code, i.e. `Malaga` vs. `MA`.

*Most importantly*, card verification via 3DS flow is being ignored. It was assumed that the `nonce` would not be given when card was invalid but this wasn't true, there is always a nonce. It's the shift in liability what has to be checked, and the status of the payload returned by `verifyCard`.

This PR addresses these problems and other potential pitfalls by fixing, improving and refactoring out `BraintreeForm` component used in the dev portal.

### Verification
Testing credit cards: https://developer.paypal.com/braintree/docs/guides/3d-secure/testing-go-live#testing

Assuming the payment gateway is set up as Braintree (sandbox account available in Bitwarden), setting `Finance` is enabled for the provider account and dev portal has feature `Finance` visible:

* For provider:
  1. Access your provider's payment details at `Settings > Billing`
  2. Add or edit the card card details

* For dev-portal:
  1. Mark *3D Secure enabled* at `Audience > Billing > Payment Gateway`
  2. Log in to dev portal go to `Settings > Credit card details > Add/Edit credit card details and billing address`
  3. Try to break the form in any way possible. The CC should only be saved if it passes 3DS verification. This modal should be required when a challenge is requested by the bank:
    Screenshot 2023-02-17 at 08 49 20

### Notes
Original PR:
* https://github.com/3scale/porta/pull/3079

### CODE REVIEW

**Comment by josemigallas** [comment]
For older discussions, see original PR: https://github.com/3scale/porta/pull/3079

**Comment by mayorova** [comment]
I noticed a few issues on this branch.

1. Maybe not important, but I don't see this happening in production - there are some stripey requests in the Network tab, any idea why?
Screenshot from 2023-02-16 11-58-49

2. With 3D secure on I am failing to add credit card details (without 3Ds works OK)

Getting a 422, with error:
```
{"error":{"message":"Billing country code format is invalid."},"threeDSecureInfo":{"liabilityShifted":false,"liabilityShiftPossible":false}}
```

The form data:
Screenshot from 2023-02-16 11-12-48


```
{
  "amount": "0.0",
  "additionalInfo": {
    "billingLine1": "street",
    "billingCity": "barcelona",
    "billingState": "",
    "billingPostalCode": "08013",
    "billingCountryCode": "Spain",
    "billingPhoneNumber": "123456",
    "billingGivenName": "Ddd",
    "billingSurname": "Mmmm"
  },
  "challengeRequested": true,
  "bin": "555555",
  "dfReferenceId": "1_272c565e-7036-455c-99b6-68a58cd27a80",
  "clientMetadata": {
    "requestedThreeDSecureVersion": "2",
    "sdkVersion": "web/3.75.0",
    "cardinalDeviceDataCollectionTimeElapsed": 155,
    "issuerDeviceDataCollectionTimeElapsed": 317,
    "issuerDeviceDataCollectionResult": true
  },
  "authorizationFingerprint": "eyJ0eXAiOiJKV1...",
  "braintreeLibraryVersion": "braintree/web/3.75.0",
  "_meta": {
    "merchantAppId": "provider.3scale.localhost:3000",
    "platform": "web",
    "sdkVersion": "3.75.0",
    "source": "client",
    "integration": "custom",
    "integrationType": "custom",
    "sessionId": "7629cce5-c995-4252-abcf-feb827885153"
  }
}
```

**Comment by josemigallas** [comment]
@mayorova 
1. Those happen because in dev portal both braintree and stripe packages are downloaded, and there are some rogue calls to stripe even though it's not enabled. I think it should happen in production as well.

2. Good catch, the country code is only set after selecting an option from the country select. I will fix it so that it also work for "preselected" values.

**Comment by mayorova** [comment]
@josemigallas now the issue I mentioned in https://github.com/3scale/porta/pull/3214#issuecomment-1432934210 is no longer reproduced.

And the card gets added successfully.

The problem is... it always gets added successfully, even for the card which I assume is supposed to fail? `3DS unsuccessful with challenge | 4000000000001109`.

Or, if on the pop-up I cancel the challenge, the card is added anyway.

Or is that expected? :thinking:

**Comment by josemigallas** [comment]
> @josemigallas now the issue I mentioned in [#3214 (comment)](https://github.com/3scale/porta/pull/3214#issuecomment-1432934210) is no longer reproduced.
> 
> And the card gets added successfully.
> 
> The problem is... it always gets added successfully, even for the card which I assume is supposed to fail? `3DS unsuccessful with challenge | 4000000000001109`.
> 
> Or, if on the pop-up I cancel the challenge, the card is added anyway.
> 
> Or is that expected? 🤔

Not at all. I missed a very important part of the verification. It should be fixed now and the card will only be stored if the 3DS flow finishes successfully.

**Review by mayorova — DISMISSED** [review]
Looks good now :+1:

File: app/javascript/src/PaymentGateways/braintree/BraintreeForm.tsx [hunk 1]
```diff
@@ -1,156 +1,254 @@
...
-    void getHostedFieldsInstance()
-  }, [])
+  const billingAddressErrors = validateForm(billingAddress)
+  const submitDisabled = (hostedFields === undefined) || !valid || (billingAddressErrors !== undefined) || loading || submitting
 
   useEffect(() => {
-    if (braintreeNonceValue && formRef.current) {
-      formRef.current.submit()
+    if (hostedFieldsError) {
+      // TODO: do something
```
Comment (jlledom) [inline app/javascript/src/PaymentGateways/braintree/BraintreeForm.tsx:31]: Why not do this?

**Review by jlledom — APPROVED** [review]
It seems to work for me

File: app/javascript/src/PaymentGateways/braintree/BraintreeForm.tsx (hunk 1 above)
Comment (josemigallas) [reply app/javascript/src/PaymentGateways/braintree/BraintreeForm.tsx:31]: Yeah but, what should we do? 😄

Comment (jlledom) [reply app/javascript/src/PaymentGateways/braintree/BraintreeForm.tsx:31]: According to the comment: "something"

**Review by mayorova — APPROVED** [review]

**Review by jlledom — APPROVED** [review]

**Pull request merged**